crates/ark/src/plots/graphics_device.rs

```diff
@@ -229,41 +340,48 @@ impl DeviceContext {
         }
     }
 
-    fn process_changes(
-        &self,
-        comm_manager_tx: Sender<CommManagerEvent>,
-        iopub_tx: Sender<IOPubMessage>,
-        dynamic_plots: bool,
-    ) {
-        // Refcell Safety: Short borrows in the file.
-        let id = unwrap!(self._id.borrow().clone(), None => {
-            log::error!("Unexpected uninitialized `id`.");
+    #[tracing::instrument(level = "trace", skip_all)]
+    fn process_changes(&self) {
+        let id = self.id();
+
+        if !self.has_changes.replace(false) {
+            log::trace!("No changes to process for plot `id` {id}");
             return;
-        });
+        }
 
-        let new_page = self._new_page.replace(false);
-        if new_page {
-            self.process_new_plot(id.as_str(), comm_manager_tx, iopub_tx, dynamic_plots);
+        log::trace!("Processing changes for plot `id` {id}");
+
+        // Record the changes so we can replay them when Positron asks us for them.
+        // Recording here overrides an existing recording for `id` if something has
+        // changed between then and now, which is what we want, for example, we want
+        // it when running this line by line:
+        //
+        // ```r
+        // par(mfrow = c(2, 1))
+        // plot(1) # Should get recorded with `id1`
+        // plot(2) # Should record and overwrite `id1` because no new_page has been requested
+        // ```
+        Self::record_plot(&id);
+
+        if self.is_new_page.replace(false) {
+            self.process_new_plot(&id);
         } else {
-            self.process_update_plot(id.as_str(), iopub_tx, dynamic_plots);
+            self.process_update_plot(&id);
         }
     }
 
-    fn process_new_plot(
-        &self,
-        id: &str,
-        comm_manager_tx: Sender<CommManagerEvent>,
-        iopub_tx: Sender<IOPubMessage>,
-        dynamic_plots: bool,
-    ) {
-        if dynamic_plots {
-            self.process_new_plot_positron(id, comm_manager_tx);
+    fn process_new_plot(&self, id: &PlotId) {
+        if self.should_use_dynamic_plots() {
+            self.process_new_plot_positron(id);
         } else {
-            self.process_new_plot_jupyter_protocol(id, iopub_tx);
+            self.process_new_plot_jupyter_protocol(id);
         }
     }
 
-    fn process_new_plot_positron(&self, id: &str, comm_manager_tx: Sender<CommManagerEvent>) {
+    #[tracing::instrument(level = "trace", skip_all, fields(id = %id))]
+    fn process_new_plot_positron(&self, id: &PlotId) {
```

Comment:
I've added tracing instrumentation everywhere, which really helped me understand what the heck was going on

Reply:
Cool